src/ngraph/runtime/gpu/cuda_emitter.cpp

```diff
@@ -676,6 +680,236 @@ size_t runtime::gpu::CUDAEmitter::build_elementwise_n_to_1(const GPURuntimeConte
+                    writer << "data_idx++;\n";
+                }
+                writer.block_end();
+                writer << "dimensions[threadIdx.x + (rank-1) * " << nthreads_per_block
+                       << "] = dim_product;\n";
+                writer << "data_idx = 0;\n";
+                writer << "bool in_bounds = true;\n";
+                writer << "int source_idx = 0;\n";
+                writer << "for (int i = threadIdx.x; i < rank * " << nthreads_per_block
+                       << "; i += " << nthreads_per_block << ")\n";
+                writer.block_begin();
+                {
+                    writer << "int source_di = division_by_invariant_multiplication(dimensions[i], "
+                              "slice_magic[data_idx], "
+                              "slice_shift[data_idx]);\n";
+                    writer << "bool on_stride = (mod16(dimensions[i], source_di, "
+                              "slice_str[data_idx]) == 0);\n";
+                    // within slice of input tensor and a multiple of the slice stride
+                    writer << "bool in_slice_di = (dimensions[i] >= lower_bounds[data_idx]) && "
+                              "(dimensions[i] < upper_bounds[data_idx]) && on_stride;\n";
+                    writer << "in_bounds = in_bounds && in_slice_di;\n";
+                    // subtract off lower bound to convert to source index
+                    writer << "source_di -= lower_bounds[data_idx];\n";
+                    writer << "source_idx += source_di * src_strides[data_idx];\n";
+                    writer << "data_idx++;\n";
+                }
+                writer.block_end();
+                writer << "out[tid] = in_bounds ? source[source_idx] : in[tid];\n";
+            }
+            writer.block_end();
+        }
+        writer.block_end();
+        compiled_kernel = ctx->compiled_kernel_pool->set(kernel_name, writer.get_code());
+    }
+
+    std::vector<int> params;
+    params.reserve(tensor_shape.size() * 3 - 1 + lower_bounds.size() + upper_bounds.size() +
+                   slice_strides.size() * 3 + source_shape.size() * 2);
+
+    int dstr_offset = params.size();
+    auto dim_strides = row_major_strides(tensor_shape);
+    for (int i = 0; i < dim_strides.size() - 1; i++)
+    {
+        params.push_back(static_cast<int>(dim_strides[i]));
+    }
+    int dmagic_offset = params.size();
+    std::vector<int> dmagics;
+    std::vector<int> dshifts;
+    for (int i = 0; i < tensor_shape.size(); i++)
+    {
+        int magic;
+        int shift;
+        std::tie(magic, shift) = idiv_magic_u64(static_cast<int>(dim_strides[i]));
+        dmagics.push_back(magic);
+        dshifts.push_back(shift);
+    }
+    std::copy(dmagics.begin(), dmagics.end(), std::back_inserter(params));
+    int dshift_offset = params.size();
+    std::copy(dshifts.begin(), dshifts.end(), std::back_inserter(params));
+    int lbound_offset = params.size();
+    for (auto const& lbound : lower_bounds)
```

Comment:
The more I see `auto` the more I don't like it. Reviewing this code I do not know what type `lbound` is. I can see it comes out of a Coordinate but now I have to go and find that class to find out what the Coordinate iterator outputs. Sigh. I think it is a size_t but I could be wrong. If it is a size_t then why not just put `size_t lbound` which is shorter and clearer. To me auto is for things like opaque handles or things you  simply pass around. Here you are casting it so you must really know what it is.

Reply:
I hear you, the reason I prefer auto for iterators is because if at some point we decide to change class Coordinate so that it inherits from std::vector<uint32_t> instead of std::vector<size_t>, then this iterator will still work and won't force an implicit cast that requires us to update this code.

Sidebar: It's a different discussion but the fact that Shape/Coordinate/Stride are all inheriting from std::vector<size_t> actually causes some pain points because size_t is 64bit wide. The fact that it is 64bit wide causes us to need to lop off the high bits with a static_cast all over in cudnn and cuda, since 64bit is much slower on the GPU and the API's require it for cudnn.

Will we ever have a tensor that has a size or stride that is larger than 2^32 ? That seems very unlikely.